packages/grafana-ui/src/components/Combobox/useOptions.ts

```diff
@@ -92,43 +92,64 @@ export function useOptions<T extends string | number>(rawOptions: AsyncOptions<T
     [debouncedLoadOptions, isAsync]
   );
 
-  const organizeOptionsByGroup = useCallback((options: Array<ComboboxOption<T>>) => {
-    const groupedOptions = new Map<string | undefined, Array<ComboboxOption<T>>>();
-    for (const option of options) {
-      const groupExists = groupedOptions.has(option.group);
-      if (groupExists) {
-        groupedOptions.get(option.group)?.push(option);
-      } else {
-        groupedOptions.set(option.group, [option]);
-      }
-    }
+  const stringifiedOptions = useMemo(() => {
+    return isAsync ? [] : rawOptions.map(itemToString);
+  }, [isAsync, rawOptions]);
 
-    // Reorganize options to have groups first, then undefined group
-    const reorganizeOptions = [];
-    for (const [group, groupOptions] of groupedOptions) {
-      if (!group) {
-        continue;
-      }
-      reorganizeOptions.push(...groupOptions);
+  // Create a list of options filtered by the current search.
+  // If async, just returns the async options.
+  const filteredOptions = useMemo(() => {
+    if (isAsync) {
+      return asyncOptions;
     }
 
-    const undefinedGroupOptions = groupedOptions.get(undefined);
-    if (undefinedGroupOptions) {
-      reorganizeOptions.push(...undefinedGroupOptions);
-    }
-    return reorganizeOptions;
-  }, []);
+    return fuzzyFind(rawOptions, stringifiedOptions, userTypedSearch);
+  }, [asyncOptions, isAsync, rawOptions, stringifiedOptions, userTypedSearch]);
 
-  const stringifiedOptions = useMemo(() => {
-    return isAsync ? [] : rawOptions.map(itemToString);
-  }, [isAsync, rawOptions]);
+  const [finalOptions, groupStartIndices] = useMemo(() => {
```

Comment:
This diff is a bit messy but I essentially:

 - moved `organizeOptionsByGroup` to module-level function `sortByGroup` outside of the hook
 - calling `sortByGroup` is now done in a seperate useMemo so it's not called on every single search.